types/json/binary_functions.go

```diff
@@ -72,6 +76,71 @@ func (bj BinaryJSON) Unquote() (string, error) {
 	}
 }
 
+// unquoteString recognizes the escape sequences shown in:
+// https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#json-unquote-character-escape-sequences
+func unquoteString(s string) (string, error) {
+	ret := new(bytes.Buffer)
+	for i := 0; i < len(s); i++ {
+		if s[i] == '\\' {
+			i++
+			if i == len(s) {
+				return "", errors.New("Missing a closing quotation mark in string")
+			}
+			switch s[i] {
+			case '"':
+				ret.WriteByte('"')
+			case 'b':
+				ret.WriteByte('\b')
+			case 'f':
+				ret.WriteByte('\f')
+			case 'n':
+				ret.WriteByte('\n')
+			case 'r':
+				ret.WriteByte('\r')
+			case 't':
+				ret.WriteByte('\t')
+			case '\\':
+				ret.WriteByte('\\')
+			case 'u':
+				if i+5 > len(s) {
+					return "", errors.Errorf("Invalid unicode: %s", s[i+1:])
+				}
+				char, size, err := decodeEscapedUnicode(hack.Slice(s[i+1 : i+5]))
+				if err != nil {
+					return "", errors.Trace(err)
+				}
+				ret.Write(char[0:size])
+				i += 5
+			default:
+				// For all other escape sequences, backslash is ignored.
+				ret.WriteByte(s[i])
+			}
+		} else {
+			ret.WriteByte(s[i])
+		}
+	}
+	return ret.String(), nil
+}
+
+// decodeEscapedUnicode decodes unicode into utf8 bytes specified in RFC 3629.
+// According RFC 3629, the max length of utf8 characters is 4 bytes.
+// And MySQL use 4 bytes to represent the unicode which must be in [0, 65536).
+func decodeEscapedUnicode(s []byte) (char [4]byte, size int, err error) {
+	size, err = hex.Decode(char[0:2], s)
+	if err != nil || size != 2 {
+		// The unicode must can be represented in 2 bytes.
+		return char, 0, errors.Trace(err)
+	}
+	var unicode uint16
+	err = binary.Read(bytes.NewReader(char[0:2]), binary.BigEndian, &unicode)
```

Comment:
I know it is an old JSON implement. I just want to know is there a test case cover this?

Reply:
There are `JSONUnquote` tests in this package and `expression`.